Add first custom checks and tests

deanmalan opened the pull request.

deanmalan committed:
Add budget-phase-required-values custom check

deanmalan commented on data_checks/checks/epre/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from .budget_phase_required_values import BudgetPhaseRequiredValues
+
+
+checks = ['budget-phase-required-values']
```
Pi/JD, I'm a bit uncertain about this. Not sure where to store a list with the available checks for a dataset type. Do you think this is okay?

jbothma replied:
ja this looks like a good place at the moment

deanmalan commented on bin/run-checks.py:
```diff
@@ -1,11 +1,30 @@
+
+    # Get the custom checks for the dataset type
+    try:
+        data_checks = __import__(f'data_checks.checks.{dataset_type}')
```
I see `importlib.import_module` is favoured over `__import__`. Will change this now.

deanmalan committed:
Use import_module

jbothma commented:
Looks good so far @delenamalan

deanmalan committed:
Add docstrings for budget phase check

jbothma commented:
It's not working for me right now - it's saying there are errors in the 2018 doc which I'm pretty sure is fine, but a minimal integration test with a local csv file would make it easier to verify.

I reckon to be able to do integration/end-to-end tests we could let the check script take a path as an argument then we can check datapackages under specified directories to verify specific outcomes. We'd probably need to move the current datapackates directory tree into a directory iike `datapackages` so that error test examples don't get run as part of the usual check run

deanmalan commented:
@jbothma Good idea. Will work on an integration test now.

deanmalan committed:
Add messages to test assertions

deanmalan committed:
Move data package files to separate directory

deanmalan committed:
Use header instead of field in budget_phase check

deanmalan committed:
Only use datapackages in datapackages dir

deanmalan committed:
Restructure and add first integration test

Move the `run_checks` method to the `data_checks` package so that we can test it.

Add test datapackages and test them.

deanmalan committed:
Add docstrings, fix imports, etc.

deanmalan commented:
@jbothma It should work now :)

deanmalan committed:
Run tests travis

deanmalan committed:
Rename test data file

deanmalan committed:
Add Budget Phase check

deanmalan committed:
Add department names check

deanmalan committed:
Columns can be in any order

deanmalan committed:
Fix pep8 errors

deanmalan committed:
Add correct url to data back in datapackage

Commit:
Update __init__.py

jbothma commented on tests/datapackages/2018-19/epre/epre.json:
```diff
@@ -0,0 +1,116 @@
+{
```
I meant we should have tests that run against the [actual schema](https://raw.githubusercontent.com/vulekamali/data-checks/master/schema/epre.json), so the test datapackages should refer to the actual schema

deanmalan replied:
@jbothma the problem is that the datapackage `.json` file must be in the same directory or a higher-directory level than the schema. So the path to the schema can't be, for example, `../../schema/epre.json`. I'm looking at whether it would be possible to create a symbolic link to the schema and then use the path to the symbolic link.

deanmalan replied:
I'll just use the url to schema :)

jbothma replied:
yeah - it's a bit of a pain - that also makes it hard to test out a change in the schema - you end up having to modify the branch name in the URLs.

Let's keep this in mind.

jbothma commented on tests/test_check_datapackage.py:
```diff
@@ -0,0 +1,43 @@
+import unittest
```
let's call this test_check_schema since the idea is for it to check the schema, rather than the datapackage.

jbothma commented on tests/test_check_datapackage.py:
```diff
@@ -0,0 +1,43 @@
+        pass
+
+    def test_perfect_datapackage(self):
+        file_path = 'tests/datapackages/2018-19/epre/perfect_datapackage.json'
```
what I had in mind here was actually for the tests to be in paths like `test/epre_perfect_datapackage/2018-19/epre` so then you run `python bin/data-checks.py test/epre_perfect_datapackage` then it's easy to run negative examples cases while developing schemas. And this suite is just there to run them all and ensure they pass or fail appropriately. And the interface of `{{ financial_year}}/{{ dataset_type }}/datapackage.json` doesn't get violated

deanmalan committed:
Add directory path argument to run-checks script

deanmalan committed:
Merge branch 'delena-custom-checks' of github.com:fluenty/data-checks into delena-custom-checks

deanmalan committed:
Use url to epre.json schema

jbothma commented on data_checks/checks/epre/budget_phase_for_new_financial_year.py:
```diff
@@ -0,0 +1,60 @@
+                message=(
+                    f'Value "{new_year_values[0].get("value")}" '
+                    'in column budget_phase must be '
+                    f' must be "{self.expected_budget_phase}" when '
```
results in `must be  must be`

jbothma commented on data_checks/checks/epre/budget_phase_for_new_financial_year.py:
```diff
@@ -0,0 +1,60 @@
+            error = Error(
+                'budget-phase-for-new-financial-year',
+                message=(
+                    f'Value "{new_year_values[0].get("value")}" '
```
results in `Value "2018" in column budget_phase must be` where I'd expect `Value "Adjusted appropriation" in column budget_phase must be`

Checked by running `python bin/run-checks.py tests/epre_faulty_datapackage/`

jbothma commented on data_checks/checks/epre/budget_phase_for_new_financial_year.py:
```diff
@@ -0,0 +1,60 @@
+                    f'is "{self.new_year}" on row '
+                    f'{new_year_values[0].get("row-number")}'
+                ),
+                row_number=new_year_values[0].get("row-number"),
```
Is it easy to provide column number here? Don't worry if it's not - you have column name in the message which is nice.

deanmalan committed:
Update budget phase error

PiDelport commented on data_checks/checks/epre/budget_phase_for_new_financial_year.py:
```diff
@@ -0,0 +1,79 @@
+                    f'Empty value '
+                    'in column budget_phase must be '
+                    f'"{self.expected_budget_phase}" when '
+                    'column budget_year '
```
`budget_year` → `financial_year` (both of these)

PiDelport reviewed (commented):
Belated review!

(I have some suggestions that I'll add as a follow-up PR, too.)
